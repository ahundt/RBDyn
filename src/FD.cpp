// Copyright 2012-2017 CNRS-UM LIRMM, CNRS-AIST JRL
//
// This file is part of RBDyn.
//
// RBDyn is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// RBDyn is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with RBDyn.  If not, see <http://www.gnu.org/licenses/>.

// associated header
#include "RBDyn/FD.h"

// includes
// RBDyn
#include "RBDyn/MultiBody.h"
#include "RBDyn/MultiBodyConfig.h"

#include <iostream>

namespace rbd
{

ForwardDynamics::ForwardDynamics(const MultiBody& mb):
	H_(mb.nrDof(), mb.nrDof()),
	H_dot_(mb.nrDof(), mb.nrDof()),
	C_(mb.nrDof()),
	I_st_(mb.nrBodies()),
	I_st_dot_(mb.nrBodies()),
	X_p_c_dot_(mb.nrBodies()),
	F_(mb.nrJoints()),
	F_dot_(mb.nrJoints()),
	acc_(mb.nrBodies()),
	f_(mb.nrBodies()),
	tmpFd_(mb.nrDof()),
	dofPos_(mb.nrJoints()),
	ldlt_(mb.nrDof())
{
	int dofP = 0;
	for(int i = 0; i < mb.nrJoints(); ++i)
	{
		F_[i].resize(6, mb.joint(i).dof());
		F_dot_[i].resize(6, mb.joint(i).dof()); ///
		dofPos_[i] = dofP;
		dofP += mb.joint(i).dof();
	}
}


void ForwardDynamics::forwardDynamics(const MultiBody& mb, MultiBodyConfig& mbc)
{
	computeH(mb, mbc);
	computeC(mb, mbc);

	paramToVector(mbc.jointTorque, tmpFd_);
	ldlt_.compute(H_);
	tmpFd_ = ldlt_.solve(tmpFd_ - C_);

	vectorToParam(tmpFd_, mbc.alphaD);
}


void ForwardDynamics::computeH(const MultiBody& mb, const MultiBodyConfig& mbc)
{
	const std::vector<Body>& bodies = mb.bodies();
	const std::vector<Joint>& joints = mb.joints();
	const std::vector<int>& pred = mb.predecessors();

	H_.setZero();
	H_dot_.setZero(); /// H_dot
	
	for(std::size_t i = 0; i < bodies.size(); ++i)
	{
		I_st_[i] = bodies[i].inertia();

		I_st_dot_[i] = Eigen::Matrix<double, 6, 6>::Zero(); /// H_dot

		if(pred[i] != -1){ /// H_dot
			const sva::MotionVecd& vj_i_neg = -mbc.jointVelocity[i]; /// H_dot
			X_p_c_dot_[i] = sva::vector6ToCrossMatrix(vj_i_neg.vector()) * mbc.parentToSon[i].matrix(); /// H_dot
		} /// H_dot
	} // end for

	
	for(int i = static_cast<int>(bodies.size()) - 1; i >= 0; --i)
	{
		if(pred[i] != -1)
		{
			const sva::PTransformd& X_p_i = mbc.parentToSon[i];
			I_st_[pred[i]] += X_p_i.transMul(I_st_[i]);

			I_st_dot_[pred[i]] = I_st_dot_[pred[i]] + 
										X_p_i.matrix().transpose() * I_st_dot_[i] * X_p_i.matrix() + 
										X_p_c_dot_[i].transpose() * I_st_[i].matrix() * X_p_i.matrix() +  
										X_p_i.matrix().transpose() * I_st_[i].matrix() * X_p_c_dot_[i]; /// H_dot
		}

		for(int dof = 0; dof < joints[i].dof(); ++dof)
		{
			F_[i].col(dof).noalias() = (I_st_[i]*
					sva::MotionVecd(mbc.motionSubspace[i].col(dof))).vector();

			F_dot_[i].col(dof).noalias() = I_st_dot_[i]*mbc.motionSubspace[i]; /// H_dot
		}

		H_.block(dofPos_[i], dofPos_[i], joints[i].dof(), joints[i].dof()).noalias() =
			mbc.motionSubspace[i].transpose()*F_[i];

		H_dot_.block(dofPos_[i], dofPos_[i], joints[i].dof(), joints[i].dof()).noalias() =
			mbc.motionSubspace[i].transpose()*F_dot_[i]; /// H_dot
		
		int j = i;
		while(pred[j] != -1)
		{
			const sva::PTransformd& X_p_j = mbc.parentToSon[j];
			
			for(int dof = 0; dof < joints[i].dof(); ++dof)
			{
				

				F_dot_[i].col(dof) = X_p_j.matrix().transpose()*F_dot_[i].col(dof) +
										X_p_c_dot_[j].transpose()*sva::ForceVecd(F_[i].col(dof)).vector(); /// H_dot

				F_[i].col(dof) = X_p_j.transMul(sva::ForceVecd(F_[i].col(dof))).vector();

			}
			j = pred[j];

			if(joints[j].dof() != 0)
			{
				H_.block(dofPos_[i], dofPos_[j], joints[i].dof(), joints[j].dof()).noalias() =
					F_[i].transpose()*mbc.motionSubspace[j];

				H_.block(dofPos_[j], dofPos_[i], joints[j].dof(), joints[i].dof()).noalias() =
					H_.block(dofPos_[i], dofPos_[j], joints[i].dof(), joints[j].dof()).transpose();

				H_dot_.block(dofPos_[i], dofPos_[j], joints[i].dof(), joints[j].dof()).noalias() =
					F_dot_[i].transpose()*mbc.motionSubspace[j]; /// H_dot

				H_dot_.block(dofPos_[j], dofPos_[i], joints[j].dof(), joints[i].dof()).noalias() =
					H_dot_.block(dofPos_[i], dofPos_[j], joints[i].dof(), joints[j].dof()).transpose(); /// H_dot
			}
		} // end while
	} // end for
}


void ForwardDynamics::computeC(const MultiBody& mb, const MultiBodyConfig& mbc)
{
	const std::vector<Body>& bodies = mb.bodies();
	const std::vector<Joint>& joints = mb.joints();
	const std::vector<int>& pred = mb.predecessors();

	sva::MotionVecd a_0(Eigen::Vector3d::Zero(), mbc.gravity);

	for(std::size_t i = 0; i < bodies.size(); ++i)
	{
		const sva::PTransformd& X_p_i = mbc.parentToSon[i];

		const sva::MotionVecd& vj_i = mbc.jointVelocity[i];

		const sva::MotionVecd& vb_i = mbc.bodyVelB[i];

		if(pred[i] != -1)
			acc_[i] = X_p_i*acc_[pred[i]] + vb_i.cross(vj_i);
		else
			acc_[i] = X_p_i*a_0 + vb_i.cross(vj_i);

		f_[i] = bodies[i].inertia()*acc_[i] + vb_i.crossDual(bodies[i].inertia()*vb_i) - mbc.bodyPosW[i].dualMul(mbc.force[i]);
	}

	for(int i = static_cast<int>(bodies.size()) - 1; i >= 0; --i)
	{
		C_.segment(dofPos_[i], joints[i].dof()).noalias() = mbc.motionSubspace[i].transpose()*f_[i].vector();

		if(pred[i] != -1)
		{
			const sva::PTransformd& X_p_i = mbc.parentToSon[i];
			f_[pred[i]] += X_p_i.transMul(f_[i]);
		}
	}
}


void ForwardDynamics::sForwardDynamics(const MultiBody& mb, MultiBodyConfig& mbc)
{
	checkMatchParentToSon(mb, mbc);
	checkMatchMotionSubspace(mb, mbc);
	checkMatchJointVelocity(mb, mbc);
	checkMatchBodyVel(mb, mbc);
	checkMatchBodyPos(mb, mbc);
	checkMatchForce(mb, mbc);
	checkMatchJointTorque(mb, mbc);

	checkMatchAlphaD(mb, mbc);

	forwardDynamics(mb, mbc);
}



void ForwardDynamics::sComputeH(const MultiBody& mb, const MultiBodyConfig& mbc)
{
	checkMatchParentToSon(mb, mbc);
	checkMatchMotionSubspace(mb, mbc);

	computeH(mb, mbc);
}



void ForwardDynamics::sComputeC(const MultiBody& mb, const MultiBodyConfig& mbc)
{
	checkMatchParentToSon(mb, mbc);
	checkMatchMotionSubspace(mb, mbc);
	checkMatchJointVelocity(mb, mbc);
	checkMatchBodyVel(mb, mbc);
	checkMatchBodyPos(mb, mbc);
	checkMatchForce(mb, mbc);

	computeC(mb, mbc);
}

} // namespace rbd
