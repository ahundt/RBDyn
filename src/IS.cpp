// This file is part of RBDyn.
//
// RBDyn is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// RBDyn is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with RBDyn.  If not, see <http://www.gnu.org/licenses/>.
#include <iostream>

// associated header
#include "IS.h"

// includes
// RBDyn
#include "util.hh"
#include "MultiBody.h"
#include "MultiBodyConfig.h"
#include "FK.h"
#include "FV.h"

namespace rbd
{

InverseStatics::InverseStatics(const MultiBody& mb):
	f_(mb.nrBodies()),
  jointTorqueJacQ_(mb.nrJoints()),
  jointTorqueJacF_(mb.nrJoints())
{
	for(int i = 0; i < static_cast<int>(mb.nrJoints()); ++i)
	{
		jointTorqueJacQ_[i].resize(mb.joint(i).dof());
		jointTorqueJacF_[i].resize(mb.joint(i).dof());
    for (size_t j = 0; j < jointTorqueJacQ_[i].size(); ++j)
    {
      jointTorqueJacQ_[i][j].resize(mb.nrDof());
      std::fill(jointTorqueJacQ_[i][j].begin(), jointTorqueJacQ_[i][j].end(), 0);
    }
    for (size_t j = 0; j < jointTorqueJacF_[i].size(); ++j)
    {
      jointTorqueJacF_[i][j].resize(mb.nrBodies()*6);
      std::fill(jointTorqueJacF_[i][j].begin(), jointTorqueJacF_[i][j].end(), 0);
    }
  }
}

void InverseStatics::inverseStatics(const MultiBody& mb, MultiBodyConfig& mbc)
{
	const std::vector<Body>& bodies = mb.bodies();
	const std::vector<Joint>& joints = mb.joints();
	const std::vector<int>& pred = mb.predecessors();

	sva::MotionVecd a_0(Eigen::Vector3d::Zero(), mbc.gravity);

	for(std::size_t i = 0; i < bodies.size(); ++i)
	{
		const sva::PTransformd& X_p_i = mbc.parentToSon[i];

		if(pred[i] != -1)
			mbc.bodyAccB[i] = X_p_i*mbc.bodyAccB[pred[i]];
		else
			mbc.bodyAccB[i] = X_p_i*a_0;

		f_[i] = bodies[i].inertia()*mbc.bodyAccB[i] -
			mbc.bodyPosW[i].dualMul(mbc.force[i]);
	}


	for(int i = static_cast<int>(bodies.size()) - 1; i >= 0; --i)
	{
		for(int j = 0; j < joints[i].dof(); ++j)
		{
			mbc.jointTorque[i][j] = mbc.motionSubspace[i].col(j).transpose()*
				f_[i].vector();
		}

		if(pred[i] != -1)
		{
			const sva::PTransformd& X_p_i = mbc.parentToSon[i];
			f_[pred[i]] = f_[pred[i]] + X_p_i.transMul(f_[i]);
		}
	}
}

void InverseStatics::computeTorqueJacobian(const MultiBody& /*mb*/,
                                            MultiBodyConfig& /*mbc*/)
{
}

void InverseStatics::computeTorqueJacobianFD(const MultiBody& mb,
                                              const MultiBodyConfig& mbc,
                                              double delta)
{
  MultiBodyConfig mbcCopy(mbc);

  size_t index = 0;
  //Differentiate wrt q
  for (size_t i = 0; i < mbc.q.size(); ++i)
    for (size_t j = 0; j < mbc.q[i].size(); ++j)
    {
      mbcCopy.q[i][j] += delta;
      forwardKinematics(mb, mbcCopy);
      forwardVelocity(mb, mbcCopy);
      inverseStatics(mb, mbcCopy);
      for (size_t k = 0; k < mbc.jointTorque.size(); ++k)
        for (size_t l = 0; l < mbc.jointTorque[k].size(); ++l)
          jointTorqueJacQ_[k][l][index] = (mbcCopy.jointTorque[k][l]-mbc.jointTorque[k][l])/delta;
      mbcCopy.q[i][j] -= delta;
      index++;
    }

  //restore original value
  forwardKinematics(mb, mbcCopy);
  forwardVelocity(mb, mbcCopy);

  //Differentiate wrt force
  index = 0;
  for (size_t i = 0; i < mbc.force.size(); ++i)
  {
    for (size_t j = 0; j < 3; ++j)
    {
      mbcCopy.force[i].force()[j] += delta;
      inverseStatics(mb, mbcCopy);
      for (size_t k = 0; k < mbc.jointTorque.size(); ++k)
        for (size_t l = 0; l < mbc.jointTorque[k].size(); ++l)
        {
          double res = (mbcCopy.jointTorque[k][l]-mbc.jointTorque[k][l])/delta;
          jointTorqueJacF_[k][l][index] = res;
        }
      mbcCopy.force[i].force()[j] -= delta;
      index++;
    }
    for (size_t j = 0; j < 3; ++j)
    {
      mbcCopy.force[i].couple()[j] += delta;
      inverseStatics(mb, mbcCopy);
      for (size_t k = 0; k < mbc.jointTorque.size(); ++k)
        for (size_t l = 0; l < mbc.jointTorque[k].size(); ++l)
          jointTorqueJacF_[k][l][index] = (mbcCopy.jointTorque[k][l]-mbc.jointTorque[k][l])/delta;
      mbcCopy.force[i].couple()[j] -= delta;
      index++;
    }
  }
  //restore original value
  inverseStatics(mb, mbcCopy);
}


void InverseStatics::sInverseStatics(const MultiBody& mb, MultiBodyConfig& mbc)
{
	checkMatchAlphaD(mb, mbc);
	checkMatchForce(mb, mbc);
	checkMatchJointConf(mb, mbc);
	checkMatchJointVelocity(mb, mbc);
	checkMatchBodyPos(mb, mbc);
	checkMatchParentToSon(mb, mbc);
	checkMatchBodyVel(mb, mbc);
	checkMatchMotionSubspace(mb, mbc);

	checkMatchBodyAcc(mb, mbc);
	checkMatchJointTorque(mb, mbc);

  inverseStatics(mb, mbc);
}

} // namespace rbd
