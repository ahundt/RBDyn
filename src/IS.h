// This file is part of RBDyn.
//
// RBDyn is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// RBDyn is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with RBDyn.  If not, see <http://www.gnu.org/licenses/>.

#pragma once

// includes
// std
#include <vector>

// SpaceVecAlg
#include <SpaceVecAlg/SpaceVecAlg>

namespace rbd
{
class MultiBody;
struct MultiBodyConfig;

struct ForceWithApplicationPoint
{
  public:
    sva::ForceVecd force;
    Eigen::Vector3d pos;
}

/**
	* Inverse Dynamics algorithm.
	*/
class InverseStatics
{
public:
	InverseStatics()
	{}
	/// @param mb MultiBody associated with this algorithm.
	InverseStatics(const MultiBody& mb);

  /**
    * Compute the inverse statics.
    * @param mb MultiBody used has model.
    * @param mbc Uses force, parentToSon, bodyPosW, motionSubspace
    * and gravity.
    * Fills bodyAccB and jointTorque.
    */
        void inverseStatics(const MultiBody& mb, MultiBodyConfig& mbc);

  /**
    * NOTE: THIS CANNOT WORK PROPERLY. If a body is equipped with a force defined on a frame attached to a body,
    * that force won't follow the body during the FD computation
		* Compute the derivatives of the torques calculated by the inverse statics
    * w.r.t. q and forces
    * WARNING: This computes the derivative of the torques w.r.t some fictitious forces applied on each body at the point (0,0,0) of the reference frame.
		* Compute the derivatives of the torques calculated by the inverse statics
    * w.r.t. q and forces.
		* @param mb MultiBody used has model.
		* @param mbc Uses force, parentToSon, bodyPosW, parentToSon, motionSubspace
    * and gravity.
		* Fills jointTorqueJacQ and jointTorqueJacF.
		*/
	void computeTorqueJacobianJoint(const MultiBody& mb, MultiBodyConfig& mbc);

	/**
    * NOTE: THIS CANNOT WORK PROPERLY. If a body is equipped with a force defined on a frame attached to a body,
    * that force won't follow the body during the FD computation
		* Compute the derivatives of the torques calculated by the inverse statics
    * w.r.t. q and forces using Finite Differences.
    * WARNING: This computes the derivative of the torques w.r.t some fictitious forces applied on each body at the point (0,0,0) of the reference frame.
		* @param mb MultiBody used has model.
		* @param mbc Uses force, parentToSon, bodyPosW, parentToSon, motionSubspace
    * and gravity.
		* Fills jointTorqueJacQ and jointTorqueJacF.
		*/
	//void computeTorqueJacobianFD(const MultiBody& mb, const MultiBodyConfig& mbc, double delta = 1e-8);
	//void computeTorqueJacobianJointFD(const MultiBody& mb, const MultiBodyConfig& mbc, double delta = 1e-8);
	//void computeTorqueJacobianForceFD(const MultiBody& mb, const MultiBodyConfig& mbc, double delta = 1e-8);

	// safe version for python binding

	/** safe version of @see inverseDynamics.
		* @throw std::domain_error If mb don't match mbc.
		*/
	void sInverseStatics(const MultiBody& mb, MultiBodyConfig& mbc);

	/**
		* @brief Get the internal forces.
		* @return vector of forces transmitted from body λ(i) to body i across
		* joint i.
		*/
	const std::vector<sva::ForceVecd>& f() const{return f_;};

  const std::vector<std::vector<Eigen::VectorXd>>& jointTorqueDiff() const
  {
    return jointTorqueDiff_;
  };

  /// @brief Vector with all the individual forces applied on each body
  std::vector<std::vector<ForceWithApplicationPoint>> allForces_;

private:
	/// @brief Internal forces.
	/// f_ is the vector of forces transmitted from body λ(i) to body i across
	/// joint i.
	std::vector<sva::ForceVecd> f_;
  std::vector<Eigen::MatrixXd> df_;
  std::vector<std::vector<Eigen::VectorXd>> jointTorqueDiff_;

};

} // namespace rbd
