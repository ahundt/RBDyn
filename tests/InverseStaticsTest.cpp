// This file is part of RBDyn.
//
// RBDyn is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// RBDyn is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with RBDyn.  If not, see <http://www.gnu.org/licenses/>.

// check memory allocation in some method

// includes
// std
#include <iostream>

// boost
#define BOOST_TEST_MODULE Statics
#include <boost/test/unit_test.hpp>
#include <boost/math/constants/constants.hpp>

// SpaceVecAlg
#include <SpaceVecAlg/SpaceVecAlg>

// RBDyn
#include "FK.h"
#include "FV.h"
#include "FD.h"
#include "IS.h"
#include "Body.h"
#include "Joint.h"
#include "MultiBody.h"
#include "MultiBodyConfig.h"
#include "MultiBodyGraph.h"
#include "util.hh"
#include "fixture.hh"

// arm
#include "XXXarm.h"

namespace rbd
{

BOOST_AUTO_TEST_CASE(XXXArmTorqueJacobian)
{
  boost::shared_ptr<boost::test_tools::output_test_stream> output =
      retrievePattern("InverseStaticsTest");

	rbd::MultiBody mb;
	rbd::MultiBodyConfig mbc;
	rbd::MultiBodyGraph mbg;
	std::tie(mb, mbc, mbg) = makeXXXarm();
  rbd::InverseStatics IS(mb);
  {
    mbc.q[1][0] = 0.0;
    mbc.q[2][0] = 0.0;
    mbc.q[3][0] = 0.0;
    forwardKinematics(mb, mbc);
    forwardVelocity(mb, mbc);
    IS.inverseStatics(mb, mbc);
    IS.computeTorqueJacobianFD(mb, mbc, 1e-10);

    (*output) << "========================================" << std::endl;
    (*output) << "Results for mbc.q =" << mbc.q << std::endl;
    (*output) << "mbc.jointTorque =\n" << mbc.jointTorque << std::endl;
    for (auto e: IS.f())
      (*output) << "IS.f().vector =\n" << e.vector() << std::endl;
    (*output) << "IS.jointTorqueJacQ() =\n" << IS.jointTorqueJacQ() << std::endl;
    (*output) << "IS.jointTorqueJacF() =\n" << IS.jointTorqueJacF() << std::endl;
  }
  {
    mbc.q[1][0] = M_PI;
    mbc.q[2][0] = 0.0;
    mbc.q[3][0] = 0.0;
    forwardKinematics(mb, mbc);
    forwardVelocity(mb, mbc);
    IS.inverseStatics(mb, mbc);
    IS.computeTorqueJacobianFD(mb, mbc, 1e-10);

    (*output) << "========================================" << std::endl;
    (*output) << "Results for mbc.q =" << mbc.q << std::endl;
    (*output) << "mbc.jointTorque =\n" << mbc.jointTorque << std::endl;
    for (auto e: IS.f())
      (*output) << "IS.f().vector =\n" << e.vector() << std::endl;
    (*output) << "IS.jointTorqueJacQ() =\n" << IS.jointTorqueJacQ() << std::endl;
    (*output) << "IS.jointTorqueJacF() =\n" << IS.jointTorqueJacF() << std::endl;
  }
  {
    mbc.q[1][0] = M_PI;
    mbc.q[2][0] = M_PI/2;
    mbc.q[3][0] = 0.0;
    forwardKinematics(mb, mbc);
    forwardVelocity(mb, mbc);
    IS.inverseStatics(mb, mbc);
    IS.computeTorqueJacobianFD(mb, mbc, 1e-10);

    (*output) << "========================================" << std::endl;
    (*output) << "Results for mbc.q =" << mbc.q << std::endl;
    (*output) << "mbc.jointTorque =\n" << mbc.jointTorque << std::endl;
    for (auto e: IS.f())
      (*output) << "IS.f().vector =\n" << e.vector() << std::endl;
    (*output) << "IS.jointTorqueJacQ() =\n" << IS.jointTorqueJacQ() << std::endl;
    (*output) << "IS.jointTorqueJacF() =\n" << IS.jointTorqueJacF() << std::endl;
  }
  {
    mbc.q[1][0] = 0.4;
    mbc.q[2][0] = 0.1;
    mbc.q[3][0] = 0.2;
    forwardKinematics(mb, mbc);
    forwardVelocity(mb, mbc);
    IS.inverseStatics(mb, mbc);
    IS.computeTorqueJacobianFD(mb, mbc, 1e-10);

    (*output) << "========================================" << std::endl;
    (*output) << "Results for mbc.q =" << mbc.q << std::endl;
    (*output) << "mbc.jointTorque =\n" << mbc.jointTorque << std::endl;
    for (auto e: IS.f())
      (*output) << "IS.f().vector =\n" << e.vector() << std::endl;
    (*output) << "IS.jointTorqueJacQ() =\n" << IS.jointTorqueJacQ() << std::endl;
    (*output) << "IS.jointTorqueJacF() =\n" << IS.jointTorqueJacF() << std::endl;
  }

  std::cout << output->str() << std::endl;
  BOOST_CHECK(output->match_pattern());
} // end of namespace rbd

}

