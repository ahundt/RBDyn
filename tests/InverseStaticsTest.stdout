


Change config to [3](0,0,0)
========================================
Results for mbc.q =[ Empty vector, [ 0], [ 0], [ 0]]
mbc.jointTorque =
[ Empty vector, [ 0], [ 0], [ 0]]
IS.f().vector =
[6](0,0,0,0,39.24,0)
IS.f().vector =
[6](0,0,0,0,29.43,0)
IS.f().vector =
[6](0,0,0,0,19.62,0)
IS.f().vector =
[6](0,0,0,0,9.81,0)
IS.jointTorqueJacQ() =
[ -29, -9.8,    0]
[-9.8, -9.8,    0]
[   0,    0,    0]
IS.jointTorqueJacF() =
[ 0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0, -1,  0,  0]
[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1, -1,  0,  0,  0,  0,  1, -1,  0,  0]
[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  2, -1,  0,  0]



Change config to [3](3.14159,0,0)
========================================
Results for mbc.q =[ Empty vector, [ 3.14159], [ 0], [ 0]]
mbc.jointTorque =
[ Empty vector, [ -3.60414e-15], [ -1.20138e-15], [ 0]]
IS.f().vector =
[6](-3.60414e-15,0,0,0,39.24,7.88861e-31)
IS.f().vector =
[6](-3.60414e-15,0,0,0,-29.43,-3.60414e-15)
IS.f().vector =
[6](-1.20138e-15,0,0,0,-19.62,-2.40276e-15)
IS.f().vector =
[6](0,0,0,0,-9.81,-1.20138e-15)
IS.jointTorqueJacQ() =
[     29,     9.8, 1.2e-15]
[    9.8,     9.8, 6.2e-16]
[      0,       0,       0]
IS.jointTorqueJacF() =
[       0,        0,        0,        0,        0,        0,        0,        0,        0,       -1,        0,        0,        0,        0, -7.5e-17,       -1,        0,        0,        0,        0, -7.5e-17,       -1,        0,        0]
[       0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0, -1.2e-16,       -1,       -1,        0,        0,        0, -1.2e-16,       -1,       -1,        0,        0]
[       0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0,        0, -2.4e-16,       -2,       -1,        0,        0]



Change config to [3](0,1.5708,0)
========================================
Results for mbc.q =[ Empty vector, [ 0], [ 1.5708], [ 0]]
mbc.jointTorque =
[ Empty vector, [ -9.81], [ -9.81], [ 0]]
IS.f().vector =
[6](-9.81,0,0,0,39.24,0)
IS.f().vector =
[6](-9.81,0,0,0,29.43,0)
IS.f().vector =
[6](-9.81,0,0,0,1.20138e-15,-19.62)
IS.f().vector =
[6](0,0,0,0,6.00689e-16,-9.81)
IS.jointTorqueJacQ() =
[-20,   0,   0]
[  0,   0,   0]
[  0,   0,   0]
IS.jointTorqueJacF() =
[ 0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0, -1,  0,  0]
[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1, -1,  0,  0,  0,  0,  1, -1,  0,  0]
[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  1, -1,  0,  0]



Change config to [3](0.4,0.1,0.2)
========================================
Results for mbc.q =[ Empty vector, [ 0.4], [ 0.1], [ 0.2]]
mbc.jointTorque =
[ Empty vector, [ -12.3436], [ -4.70316], [ 0]]
IS.f().vector =
[6](-12.3436,0,0,0,39.24,1.77636e-15)
IS.f().vector =
[6](-12.3436,0,0,0,27.1068,-11.4606)
IS.f().vector =
[6](-4.70316,0,0,0,17.2182,-9.40633)
IS.f().vector =
[6](0,0,0,0,7.5031,-6.31978)
IS.jointTorqueJacQ() =
[ -27, -8.6,    0]
[-8.6, -8.6,    0]
[   0,    0,    0]
IS.jointTorqueJacF() =
[    0,     0,     0,     0,     0,     0,     0,     0,     0,    -1,     0,     0,     0,     0,     0,    -1,     0,     0,     0,     0,     0,    -1,     0,     0]
[    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0, -0.39,  0.92,    -1,     0,     0,     0, -0.39,  0.92,    -1,     0,     0]
[    0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0,     0, -0.87,   1.8,    -1,     0,     0]

